src/Numerics/DGMethods/DGFVModel_kernels.jl

```diff
@@ -0,0 +1,998 @@
+        # element
+        eV = periodicstack ? nvertelem : 1
+
+        # Figure out the data we need
+        els = ntuple(Val(stencil_diameter)) do k
+            eH + mod1(eV - 1 + k - (stencil_center - 1), nvertelem)
+        end
+
+        # Load all the stencil data
+        @unroll for k in 1:stencil_diameter
+            load_data!(
+                local_state_prognostic[k],
+                local_state_auxiliary[k],
+                local_state_gradient_flux[k],
+                els[k],
+            )
+            # If local cell weights are NOT _M we need to load _vMI out of sgeo
+            local_cell_weights[k] = vgeo[n, _M, els[k]]
+        end
+
+        # transform all the data into primitive variables
+        @unroll for k in 1:stencil_diameter
+            prognostic_to_primitive!(
+                balance_law,
+                local_state_primitive[k],
+                local_state_prognostic[k],
+                local_state_auxiliary[k],
+            )
+        end
+        vMI[2] = 1 / local_cell_weights[stencil_center]
+
+        # If we are periodic we reconstruct the top and bottom values for eV
+        # then start with eV update in loop below
+        if periodicstack
+            # Reconstruct the top and bottom values
+            rng1, rng2 = stencil_center .+ (-stencil_width, stencil_width)
+            rng = SUnitRange(rng1, rng2)
+            reconstruction!(
+                local_state_face_primitive[1],
+                local_state_face_primitive[2],
+                local_state_primitive[rng],
+                local_state_auxiliary[rng],
+                local_cell_weights[rng],
+            )
+
+            # Transform the values back to prognostic state
+            @unroll for f in 1:2
+                primitive_to_prognostic!(
+                    balance_law,
+                    local_state_face_prognostic[f],
+                    local_state_face_primitive[f],
+                    # Use the cell auxiliary data
+                    local_state_auxiliary[stencil_center],
+                )
+            end
+        else
+
+            bctag = elemtobndy[faces[1], eH + eV]
+
+            sM = sgeo[_sM, n, faces[1], eH + eV]
+            normal = SVector(
+                sgeo[_n1, n, faces[1], eH + eV],
+                sgeo[_n2, n, faces[1], eH + eV],
+                sgeo[_n3, n, faces[1], eH + eV],
+            )
+
+            # Reconstruction using only eVs cell value
+            rng = SUnitRange(stencil_center, stencil_center)
+            # Need two geopotential cell values for hydrostatic reconstuction
+            rng_aux = SUnitRange(stencil_center, stencil_center + 1)
+            reconstruction!(
+                local_state_face_primitive[1],
+                local_state_face_primitive[2],
+                local_state_primitive[rng],
+                local_state_auxiliary[rng_aux],
+                local_cell_weights[rng],
+            )
```

Comment:
we should update the reconstruction! interface by giving the balance law as the first argument, and using vertical distance instead of local_cell_weights then we can avoid this rng_aux